modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/CursorInitializationPhaseHandler.java

```diff
@@ -54,11 +55,9 @@ public Result handle(Query query) {
 
                     query.cursor = cursor;
 
-                    query.cancel.add(timeout -> {
-                        if (!timeout) {
-                            cursor.closeAsync(true);
-                        }
-                    });
+                    query.cancel.add(timeout -> dataCursor.cancelAsync(
+                            timeout ? CancellationReason.TIMEOUT : CancellationReason.CANCEL
```

Comment:
1. can we also remove the code that adds timeout handling in `executeDdl` and `executeKill` methods of  `ExecutionServiceImpl`?
2. сan we add a test that will reproduce the issue the patch solves?

Reply:
I reread the ticket description and I want to clarify that if it is difficult to write a reproducer, then you can ignore this note.

Reply:
> can we also remove the code that adds timeout handling in executeDdl and executeKill methods of ExecutionServiceImpl?

Unfortunately, this is not only about removal from these two methods. We also need to remove according tests from `ExecutionServiceImplTest` (which is ok) and write another somewhere replacing old ones (since we would like to rely on cancellation code in one of the ExecutionPhase, new tests must involve `TestCluster`). I believe this is out of the scope of current issue. Besides, we already has [IGNITE-23792](https://issues.apache.org/jira/browse/IGNITE-23792), thus I would prefer to don't touch this part in my PR, If you don't mind.

> сan we add a test that will reproduce (constantly) the issue that this patch solves?
tbh, I have no idea how to make it to _steadily_ reproduces this race. During investigation, I just added `Thread.sleep(2)` before adding cancellation handling for cursor, but this is not acceptable to have `sleep()` in the production code for this. To make sure (kinda) that problem is gone, I've run JDBC job on TC 15 times and got no failures. For my previous patch tests had failed after ~5th iterations.